gobblin-hive-registration/src/main/java/gobblin/hive/HiveRegister.java

```diff
@@ -0,0 +1,413 @@
+   */
+  public static void register(Iterable<String> paths, State state) throws IOException {
+    try (HiveRegister hiveRegister = new HiveRegister(state)) {
+      HiveRegistrationPolicy policy = HiveRegistrationPolicyBase.getPolicy(state);
+      for (String path : paths) {
+        hiveRegister.register(policy.getHiveSpec(new Path(path)));
+      }
+    }
+  }
+
+  private boolean evaluatePredicates(HiveSpecWithPredicates spec) {
+    for (Predicate<HiveRegister> pred : spec.getPredicates()) {
+      if (!pred.apply(this)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  private void createDbIfNotExists(HiveMetaStoreClient client, String dbName) {
+    Database db = new Database();
+    db.setName(dbName);
+
+    Preconditions.checkState(this.hiveDbRootDir.isPresent(),
+        "Missing required property " + HiveRegProps.HIVE_DB_ROOT_DIR);
+    db.setLocationUri(new Path(this.hiveDbRootDir.get(), dbName + HIVE_DB_EXTENSION).toString());
+
+    HiveLock lock = locks.getDbLock(dbName);
+    lock.lock();
+
+    try {
+      client.createDatabase(db);
+      log.info("Created database " + dbName);
+    } catch (AlreadyExistsException e) {
+      // Database already exists. Nothing to do.
+    } catch (TException e) {
+      log.error("Unable to create Hive database " + dbName);
+      throw Throwables.propagate(e);
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  private Table createOrAlterTable(HiveMetaStoreClient client, String dbName, String tableName,
+      Optional<HivePartition> partition, StorageDescriptor sd) {
+
+    Table newTable = createTableWithoutRegistering(dbName, tableName, partition, sd);
+
+    HiveLock lock = locks.getTableLock(dbName, tableName);
+    lock.lock();
+
+    try {
+      if (client.tableExists(dbName, tableName)) {
+        Table existingTable = client.getTable(dbName, tableName);
+        if (needToUpdateTable(existingTable, newTable)) {
+          client.alter_table(dbName, tableName, newTable);
+          log.info(String.format("updated Hive table %s in db %s", tableName, dbName));
+          return newTable;
+        } else {
+          log.info(String.format("Hive table %s in db %s exists and no need to update", tableName, dbName));
+          return existingTable;
+        }
+      } else {
+        client.createTable(newTable);
+        log.info(String.format("Created Hive table %s in db %s", tableName, dbName));
+        return newTable;
+      }
+    } catch (TException e) {
+      log.error(String.format("Error in creating or altering Hive table %s in db %s", tableName, dbName));
+      throw Throwables.propagate(e);
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  private Table createTableWithoutRegistering(String dbName, String tableName, Optional<HivePartition> partition,
+      StorageDescriptor sd) {
+    Map<String, String> parameters = Maps.newHashMap();
+    parameters.put("EXTERNAL", Boolean.TRUE.toString());
```

Comment:
Is this needed considering the `table.setTableType(TableType.EXTERNAL_TABLE.toString())` is also invoked.